atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -0,0 +1,816 @@
+                                    final ColumnSelection columnSelection,
+                                    final long timestamp) {
+        return runRead(tableRef, new Function<com.palantir.atlasdb.keyvalue.dbkvs.impl.DbReadTable, Map<Cell, Value>>() {
+            @Override
+            public Map<Cell, Value> apply(DbReadTable table) {
+                ClosableIterator<AgnosticLightResultRow> iter = table.getLatestRows(rows, columnSelection, timestamp, true);
+                return extractResults(table, iter);
+            }
+        });
+    }
+
+    @Override
+    public Map<Cell, Value> get(TableReference tableRef, final Map<Cell, Long> timestampByCell) {
+        return runRead(tableRef, new Function<DbReadTable, Map<Cell, Value>>() {
+            @Override
+            public Map<Cell, Value> apply(DbReadTable table) {
+                ClosableIterator<AgnosticLightResultRow> iter = table.getLatestCells(timestampByCell, true);
+                return extractResults(table, iter);
+            }
+        });
+    }
+
+    @SuppressWarnings("deprecation")
+    private Map<Cell, Value> extractResults(DbReadTable table, ClosableIterator<AgnosticLightResultRow> iter) {
+        boolean hasOverflow = table.hasOverflowValues();
+        Map<Cell, Value> results = Maps.newHashMap();
+        Map<Cell, OverflowValue> overflowResults = Maps.newHashMap();
+        try {
+            while (iter.hasNext()) {
+                AgnosticLightResultRow row = iter.next();
+                Cell cell = Cell.create(row.getBytes("row_name"), row.getBytes("col_name"));
+                Long overflowId = hasOverflow ? row.getLongObject("overflow") : null;
+                if (overflowId == null) {
+                    Value value = Value.create(row.getBytes("val"), row.getLong("ts"));
+                    Value oldValue = results.put(cell, value);
+                    if (oldValue != null && oldValue.getTimestamp() > value.getTimestamp()) {
+                        results.put(cell, oldValue);
+                    }
+                } else {
+                    OverflowValue ov = new OverflowValue(row.getLong("ts"), overflowId);
+                    OverflowValue oldOv = overflowResults.put(cell, ov);
+                    if (oldOv != null && oldOv.ts > ov.ts) {
+                        overflowResults.put(cell, oldOv);
+                    }
+                }
+            }
+        } finally {
+            iter.close();
+        }
+        for (Iterator<Entry<Cell, OverflowValue>> entryIter = overflowResults.entrySet().iterator(); entryIter.hasNext(); ) {
+            Entry<Cell, OverflowValue> entry = entryIter.next();
+            Value value = results.get(entry.getKey());
+            if (value != null && value.getTimestamp() > entry.getValue().ts) {
+                entryIter.remove();
+            }
+        }
+        if (!overflowResults.isEmpty()) {
+            Map<Long, byte[]> values = Maps.newHashMapWithExpectedSize(overflowResults.size());
```

Comment:
Extract this overflowResults processing to a separate method as this is uncommon case, and we want the above code JITed & hot